### PR TITLE
Normalize subcomponents schema

### PR DESCRIPTION
**Pull request by johnclary**
## Associated issues
- https://github.com/cityofaustin/atd-data-tech/issues/11452

This PR adds a many <> many table between `moped_components` and `moped_subcomponents`. We will be adding many more subcomponents in the near future, so it seemed worthwhile to get this taken care of.

## Testing
**URL to test:** Local

**Steps to test:**
1. Create new components with subcomponents. Signals of subtype PHB, Traffic, and RRFB  should have three subcomponent options: Audible push button, Bicycle signal, Leading pedestrian interval. Bike lane components (except **Ramp to Sidewalk** and **Advisory**) have Raised pavement markers.

2. Edit components by adding and removing subcomponents.

3. If you're feeling thorough, you can run this against a copy of production. I have done this and the data looks goood.

4. You can peep the components data dictionary (https://localhost:3000/moped/dev/lookups) —it should match staging:

Component ID | Component name | Component subtype | Geometry type | Subcomponents
-- | -- | -- | -- | --
5 | Bike Box |   | Point |  
7 | Bike Lane | Buffered | Line | Raised pavement markers
8 | Bike Lane | Climbing | Line | Raised pavement markers
9 | Bike Lane | Colored | Line | Raised pavement markers
10 | Bike Lane | Contraflow | Line | Raised pavement markers
11 | Bike Lane | Protected One-Way | Line | Raised pavement markers
12 | Bike Lane | Protected Two-Way | Line | Raised pavement markers
21 | Bike Lane | Turn Lane | Line | Raised pavement markers
22 | Bike Lane | Wide Curb Lane | Line | Raised pavement markers
6 | Bike Lane |   | Line | Raised pavement markers
20 | Bike Parking | Corral | Point |  
19 | Bike Parking |   | Point |  
16 | Signal | PHB | Point | Audible push button, Bicycle signal, Leading pedestrian interval
17 | Signal | RRFB | Point | Audible push button, Bicycle signal, Leading pedestrian interval
18 | Signal | Traffic | Point | Audible push button, Bicycle signal, Leading pedestrian interval

---
#### Ship list
- [ ] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

### CODE REVIEW

File: moped-database/metadata/tables.yaml
Comment (johnclary) [inline moped-database/metadata/tables.yaml:893]: chipping away at:
- https://github.com/cityofaustin/atd-data-tech/issues/10525

File: moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql
Comment (johnclary) [inline moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql:54]: 😵 
we're going from a table that looks like this:

component_id | subcomponent_id | subcomponent_name
-- | -- | --
16 | 1 | Audible push button'
16 | 2 | Bicycle signal'
16 | 3 | Leading pedestrian interval'
17 | 4 | Audible push button'
17 | 5 | Bicycle signal'
17 | 6 | Leading pedestrian interval'
18 | 7 | Audible push button'
18 | 8 | Bicycle signal'
18 | 9 | Leading pedestrian interval'
7 | 10 | Raised pavement markers'
8 | 11 | Raised pavement markers'
9 | 12 | Raised pavement markers'
10 | 13 | Raised pavement markers'
11 | 14 | Raised pavement markers'
12 | 15 | Raised pavement markers'
6 | 16 | Raised pavement markers'
21 | 17 | Raised pavement markers'
22 | 18 | Raised pavement markers'

to a table that looks like this:


subcomponent_id | subcomponent_name
-- | --
1 | Audible push button'
2 | Bicycle signal'
3 | Leading pedestrian interval'
4 | Raised pavement markers'

File: moped-editor/src/queries/components.js
Comment (johnclary) [inline moped-editor/src/queries/components.js:50]: this was kinda tripping me up, b/c we were sourcing the subcomponent options from either the proj_component or the lookup table depending on edit state. so the changes to `useSubcomponentOptions` signature below are following from this change to always just source the allowed subcomponents from the lookup table data.

File: moped-editor/src/queries/tableLookups.js
Comment (johnclary) [inline moped-editor/src/queries/tableLookups.js:28]: fixes the random sorting currently in prod.

File: moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql
Comment (johnclary) [inline moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql:1]: 😵 
we're going from a table that looks like this:

component_id | subcomponent_id | subcomponent_name
-- | -- | --
16 | 1 | Audible push button'
16 | 2 | Bicycle signal'
16 | 3 | Leading pedestrian interval'
17 | 4 | Audible push button'
17 | 5 | Bicycle signal'
17 | 6 | Leading pedestrian interval'
18 | 7 | Audible push button'
18 | 8 | Bicycle signal'
18 | 9 | Leading pedestrian interval'
7 | 10 | Raised pavement markers'
8 | 11 | Raised pavement markers'
9 | 12 | Raised pavement markers'
10 | 13 | Raised pavement markers'
11 | 14 | Raised pavement markers'
12 | 15 | Raised pavement markers'
6 | 16 | Raised pavement markers'
21 | 17 | Raised pavement markers'
22 | 18 | Raised pavement markers'

to a table that looks like this:


subcomponent_id | subcomponent_name
-- | --
1 | Audible push button'
2 | Bicycle signal'
3 | Leading pedestrian interval'
4 | Raised pavement markers'

File: moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql
Comment (johnclary) [inline moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql:15]: we do not have event triggers configured for this table. if we did, we would want to strongly consider disabling the trigger during these bulk updates.

File: moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql
Comment (johnclary) [inline moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql:27]: we do not have event triggers configured for this table. if we did, we would want to strongly consider disabling the trigger during these bulk updates.

File: moped-editor/src/queries/tableLookups.js
Comment (frankhereford) [reply moped-editor/src/queries/tableLookups.js:28]: Thank you! Consistency in the app interface from visit to visit is quality polish.

**Review by frankhereford — APPROVED** [review]
Thank you John! This looks great to me - I've reviewed the code and also tested the app with your instructions. I didn't pull a copy of production to run it though, but thank you for doing that. 🚀

File: moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql
Comment (mddilley) [reply moped-database/migrations/1676056649901_dedupe_subcomponents_table/up.sql:1]: Woohoo! 🚀

File: moped-editor/src/queries/tableLookups.js
Comment (mddilley) [reply moped-editor/src/queries/tableLookups.js:28]: ✨

File: moped-editor/src/queries/components.js
Comment (mddilley) [reply moped-editor/src/queries/components.js:50]: Thanks for catching this! 🙏

**Review by mddilley — APPROVED** [review]
Awesome! I checked this out with the prod data locally and 👀'd at the lookups UI as well, and everything is great. Thanks for putting in the time to get these relationships sorted out (and adding some nice details to the code while you were at it!). 🛥️💨

**Comment by mddilley** [comment]
@johnclary I caught a couple of unrelated bugs while I was testing:
- When you update subcomponents on a signal component before the async request for the signals finishes, the form doesn’t register the update and the save button is disabled even though the number of selected subcomponents has changed
- Selected subcomponents don’t clear out if you change the component type. New Component > Select Bike Lane - Climbing > Select Raised pavement markers > Change Component Type to Signal - PHB > Raised pavement markers is still selected even though that isn’t a valid option for Signal - PHB subcomponents

The second one is more urgent since I was able to save the component with a non-matching subcomponent and then the UI breaks 😖

**Comment by johnclary** [comment]
@mddilley thanks for flagging those bugs. I [added a hook](https://github.com/cityofaustin/atd-moped/pull/978/commits/afdc41aaa4abef90af6220285d04a21a4f5d96c1) to reset the selected subcomponents on component change. This is not super ideal—since there are cases where we don't need to reset the subcomponents, and so it will cause user annoyance in some situations. B/c of circular dependencies, I couldn't figure out a way to do a fancy/conditional subcomponent reset without going down the rabbit hole of a reducer or custom form resolver.

**Comment by mddilley** [comment]
@johnclary Sounds great - that sounds like a good trade-off to me since these are edge cases and I think a WSOD would be more annoying than having already-chosen selections cleared. Good call because I didn't think about the form needing to be wise about what subcomponents are selected and if they match a component before clearing or not-clearing.

**Comment by johnclary** [comment]
i'm going to leave this open for now b/c @tillyw this will cause conflicts with https://github.com/cityofaustin/atd-moped/pull/984.

**Review by tillyw — COMMENTED** [review]
sorry for not reviewing before this was merged but it looks good to me - thanks for doing this!!

**Pull request merged**
